lib/ansible/parsing/vault/__init__.py

```diff
@@ -1356,11 +1355,51 @@ def _encrypt_pycrypto(b_plaintext, b_key1, b_key2, b_iv):
 
         return to_bytes(hmac.hexdigest(), errors='surrogate_or_strict'), hexlify(b_ciphertext)
 
+    @staticmethod
+    def _gen_salt(b_plaintext=None, secret=None):
+        # If no secret is provided, we should not generate the salt based on the hash.
+        # If we would do that, we leak the sha256 hash of the plaintext.
+        # This is most of the time not a problem, except if the plaintext contains guessable
+        # content or is partially known by the attacker. The attacker then could simply
+        # bruteforce the plaintext content. This is not a problem for larger files, where
+        # the structure is not known by the attacker.
+        #
+        # If we use inline encryption and only encrypt the password of the following variable
+        #   password = '1234'
+        # an attacker could take the sha256 hash out of the salt value
+        # and use it to bruteforce the password, as sha256('1234') == salt.
+        #
+        # If a secret was provided, we first append it the the plaintext and than generate the
+        # hash. By doing so we increase the length of the plaintext and also add a string that is
+        # in all cases unknown to the attacker (the passphrase).
+        # So the plaintext hash is now completely useless to the attacker.
+        #
+        # Therefore we also use a randomn salt if no plaintext is provided, or we would leak the
+        # sha256 hash of the secret.
+
+        if not secret or not b_plaintext:
+            return os.urandom(32)
+        # assert secret
+        # assert b_plaintext
+
+        b_secret = secret.bytes
+        b_plaintextAndSecret = b_plaintext + b_secret
+
+        if HAS_CRYPTOGRAPHY:
+            digest = hashes.Hash(hashes.SHA256(), backend=default_backend())
+            digest.update(b_plaintextAndSecret)
+            b_salt = digest.finalize()
+        elif HAS_PYCRYPTO:
+            b_salt = SHA256_pycrypto.new(b_plaintextAndSecret).digest()
```

Comment:
In today's meeting, @jborean93 brought up that we could use a pbkdf function here to generate the salt rather than a simple sha256sum hash.  The rationale would be if one was attempting to brute force the hash by guessing the plaintext and secret, they would have to run their guess through more hashing rounds if we use a pbkdf function than if we simply run sha256sum on the input once.

To use a pbkdf function we'd have to hardcode a salt to use with that function but I don't think that's any worse than calling sha256sum() on  b_plaintextAndSecret with no salt applied.  Correct?

Reply:
For context, this was proposed because the difference between trying to brute force the ciphertext and trying to brute-force the hash being used for the salt was brought up.  The pbkdf function making it more expensive to generate the encryption key from the password rather than generating the hash was one difference that was brought up.